File: python/matrix/matrix.py
"""Matrix exercise from exercism.io python path.

Given a string representing a matrix of numbers,
return the rows and columns of that matrix.
"""


class Matrix:
    """Matrix class

    Given a string representing a matrix of numbers,
    return the rows and columns of that matrix.
    """
    def __init__(self, matrix_string):
        self.matrix_array = self._to_array(matrix_string)

    def row(self, index):
        """Get row values by its index.

        Args:
            index: positive integer.

        Returns:
            An array of integers representing the row values.
        """
        return self.matrix_array[index - 1]

    def column(self, index):
        """Get column values by its index.

        Args:
            index: positive integer.

        Returns:
            An array of integers representing the column values.
        """
        columns = []
        for row in self.matrix_array:
            columns.append(row[index - 1])
        return columns

    @classmethod
    def _to_array(cls, matrix_string):
        """Get multi-dimensional array

        Args:
            string: string representing the matrix, e.g. "1 2 3\n4 5 6\n7 8 9"

        Returns:
            A multi-dimensional array representing the matrix.
        """
        matrix_array = []
        plain_rows = matrix_string.split('\n')
        for row in plain_rows:
            rows = []
            values = row.split(' ')
            for value in values:
                rows.append(int(value))
            matrix_array.append(rows)
        return matrix_array

File: javascript/matrix/matrix.js
export class Matrix {
  rows = [];
  columns = [];

  constructor(matrix) {
    this.rows = this._to_array(matrix);
    this.columns = this._transpose(this.rows);
  }

  get rows() {
    return this.rows;
  }

  get columns() {
    return this.columns;
  }

  _to_array(matrix) {
    return matrix
      .split("\n")
      .map((row) => row.split(" ").map((value) => parseInt(value)));
  }

  _transpose(arr) {
    return Object.keys(arr[0]).map((curr) => arr.map((row) => row[curr]));
  }
}
